released_plugins/v3d_plugins/MIND/soma_segmentation_plugin.cpp

```diff
@@ -255,14 +303,340 @@ void reconstruction_func(V3DPluginCallback2 &callback, QWidget *parent,
+  }
+
+  int seedIdx = z0 * ydim * xdim + y0 * xdim + x0;
+  unsigned char seedValue = floodData[seedIdx];
+
+  int neighborhood = 5;
+  double localMean = 0;
+  double localMax = 0;
+  double localMin = 255;
+  int count = 0;
+
+  for (int dz = -neighborhood; dz <= neighborhood; dz++)
+  {
+    for (int dy = -neighborhood; dy <= neighborhood; dy++)
+    {
+      for (int dx = -neighborhood; dx <= neighborhood; dx++)
+      {
+        int nx = x0 + dx;
+        int ny = y0 + dy;
+        int nz = z0 + dz;
+
+        if (nx >= 0 && nx < xdim && ny >= 0 && ny < ydim && nz >= 0 && nz < zdim)
+        {
+          unsigned char val = floodData[nz * ydim * xdim + ny * xdim + nx];
+          localMean += val;
+          localMax = std::max(localMax, (double)val);
+          localMin = std::min(localMin, (double)val);
+          count++;
+        }
+      }
+    }
+  }
+  localMean /= count;
+
+  printf("Local statistics: mean=%.2f, min=%.2f, max=%.2f\n", localMean, localMin, localMax);
+
+  double dynamicThreshold = std::max(threshold, static_cast<int>(marker.radius * 0.8));
+  int lowerThreshold = std::max(0, static_cast<int>(localMin - dynamicThreshold));
+  int upperThreshold = std::min(255, static_cast<int>(localMax + dynamicThreshold));
+
+  printf("Thresholds: lower=%d, upper=%d, dynamic=%.2f\n", lowerThreshold, upperThreshold, dynamicThreshold);
+
+  size_t maxRegionSize = static_cast<size_t>(4.0 / 3.0 * M_PI * pow(marker.radius * 3, 3));
+  size_t minRegionSize = 5;
+
+  std::vector<bool> visited(xdim * ydim * zdim, false);
+
+  std::queue<V3DLONG> q;
+  q.push(seedIdx);
+  visited[seedIdx] = true;
+
+  int directions[6][3] = {
+      {1, 0, 0}, {-1, 0, 0}, {0, 1, 0}, {0, -1, 0}, {0, 0, 1}, {0, 0, -1}};
+
+  min_x = x0;
+  max_x = x0;
+  min_y = y0;
+  max_y = y0;
+  min_z = z0;
+  max_z = z0;
+
+  size_t regionSize = 0;
+
+  while (!q.empty())
+  {
+    V3DLONG current = q.front();
+    q.pop();
+
+    int cz = current / (xdim * ydim);
+    int cy = (current % (xdim * ydim)) / xdim;
+    int cx = current % xdim;
+
+    double distance = std::sqrt(pow(cx - x0, 2) + pow(cy - y0, 2) + pow(cz - z0, 2));
+    if (distance > marker.radius * 4)
+      continue;
+
+    for (int i = 0; i < 6; ++i)
+    {
+      int nx = cx + directions[i][0];
+      int ny = cy + directions[i][1];
+      int nz = cz + directions[i][2];
+
+      if (nx < 0 || nx >= xdim || ny < 0 || ny >= ydim || nz < 0 || nz >= zdim)
+        continue;
+
+      V3DLONG neighborIdx = nz * ydim * xdim + ny * xdim + nx;
+
+      if (!visited[neighborIdx])
+      {
+        unsigned char neighborValue = floodData[neighborIdx];
+
+        double gradient = std::abs(static_cast<double>(neighborValue) - floodData[current]);
+
+        if (neighborValue >= lowerThreshold &&
+            neighborValue <= upperThreshold &&
+            gradient < dynamicThreshold * 2)
+        {
+
+          q.push(neighborIdx);
+          visited[neighborIdx] = true;
+          floodData[neighborIdx] = 255;
```

Comment:
This is using an 8 bit value to store floodings, but we only need binary values right? Would be much more space efficient